cpp/include/cugraph/prims/gather_edges.cuh

```diff
@@ -0,0 +1,495 @@
+     vertex_count_offsets  = vertex_count_offsets.data(),
+     global_degree_offsets = global_degree_offsets.data(),
+     graph_offsets         = graph_offsets.data(),
+     graph_indices         = graph_indices.data(),
+     sources               = sources.data(),
+     destinations          = destinations.data(),
+     dst_gpu_ids           = destination_gpu_ids.data(),
+     invalid_vertex_id,
+     indices_per_source] __device__(auto index) {
+      // source which this edge index refers to
+      auto loc           = index / indices_per_source;
+      auto source        = active_sources[loc];
+      sources[index]     = source;
+      dst_gpu_ids[index] = active_source_gpu_ids[loc];
+
+      // Find which partition id did the source belong to
+      auto partition_id =
+        thrust::distance(
+          vertex_id_seg,
+          thrust::upper_bound(
+            thrust::seq, vertex_id_seg, vertex_id_seg + vertex_id_seg_count, source)) -
+        1;
+      // starting position of the segment within global_degree_offset
+      // where the information for partition (partition_id) starts
+      //  vertex_count_offsets[partition_id]
+      // The relative location of offset information for vertex id v within
+      // the segment
+      //  v - seg[partition_id]
+      auto location_in_segment = source - vertex_id_seg[partition_id];
+
+      // csr offset value for vertex v that belongs to partition (partition_id)
+      auto csr_offset       = graph_offsets[partition_id][location_in_segment];
+      auto local_out_degree = graph_offsets[partition_id][location_in_segment + 1] - csr_offset;
+      edge_t const* adjacency_list = graph_indices[partition_id] + csr_offset;
+      // read location of global_degree_offset needs to take into account the
+      // partition offsets because it is a concatenation of all the offsets
+      // across all partitions
+      auto location        = location_in_segment + vertex_count_offsets[partition_id];
+      auto g_degree_offset = global_degree_offsets[location];
+      auto g_dst_index     = edge_index_first[index];
+      if ((g_dst_index >= g_degree_offset) && (g_dst_index < g_degree_offset + local_out_degree)) {
+        destinations[index] = adjacency_list[g_dst_index - g_degree_offset];
+      } else {
+        destinations[index] = invalid_vertex_id;
+      }
+    });
+  auto input_iter = thrust::make_zip_iterator(
+    thrust::make_tuple(sources.begin(), destinations.begin(), destination_gpu_ids.begin()));
+
+  auto compacted_length = thrust::distance(
+    input_iter,
+    thrust::remove_if(
+      handle.get_thrust_policy(),
+      input_iter,
+      input_iter + destinations.size(),
+      destinations.begin(),
+      [invalid_vertex_id] __device__(auto dst) { return (dst == invalid_vertex_id); }));
+  sources.resize(compacted_length, handle.get_stream());
+  destinations.resize(compacted_length, handle.get_stream());
+  destination_gpu_ids.resize(compacted_length, handle.get_stream());
```

Comment:
And should we really return destination_gpu_ids? We can return just the number of edges for each vertex partition.

Reply:
This is a requirement for sampling. The destination_gpu_ids are more like properties associated with each source. It has nothing to do with the gpu id of where this code is run.